File: Interview Preparation Kit - CPP/08. Dynamic Programming/001. Max Array Sum.cpp
// Problem: https://www.hackerrank.com/challenges/max-array-sum/problem
// Score: 20


#include <iostream>
#include <vector>
using namespace std;


int main(){
    int n;
    cin >> n;
    vector<int> arr(n);

    for (int i = 0; i < n; i++){
        cin >> arr[i];
    }

    vector<int> dp(n);
    dp[0] = arr[0];
    dp[1] = max(arr[0], arr[1]);
    for (int i = 2; i < n; i++){
        dp[i] = max(max(dp[i-2] + arr[i], dp[i - 1]), arr[i]);
    }

    cout << dp[n - 1];

    return 0;
}

File: Interview Preparation Kit - Python/08. Dynamic Programming/001. Max Array Sum.py
# Problem: https://www.hackerrank.com/challenges/max-array-sum/problem
#  Score: 20


def max_subset_sum(arr):
    dp = list()
    dp.append(arr[0])
    dp.append(max(arr[:2]))
    ans = dp[-1]
    for i in arr[2:]:
        dp.append(max(i, dp[-2] + i, ans))
        ans = max(ans, dp[-1])
    return ans


n = int(input())
arr = list(map(int, input().split()))
print(max_subset_sum(arr))
